### PR TITLE
Make ribs.emitters.opt public

### PR DESCRIPTION
**Pull request by btjanaka**
## TODO

- [x] Add to documentation
- [x] Rearrange files
- [x] Provide methods in GradientOptBase
- [x] Add docstrings
- [x] Fix Adam bugs
- [x] Test CMA-MEGA and CMA-MEGA (adam)

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: ribs/emitters/opt/__init__.py [hunk 1]
```diff
@@ -1,9 +1,27 @@
...
+from ribs.emitters.opt._adam_opt import AdamOpt
 from ribs.emitters.opt._cma_es import CMAEvolutionStrategy
-from ribs.emitters.opt._gradients import AdamOpt, GradientAscentOpt
+from ribs.emitters.opt._evolution_strategy_base import EvolutionStrategyBase
+from ribs.emitters.opt._gradient_ascent_opt import GradientAscentOpt
+from ribs.emitters.opt._gradient_opt_base import GradientOptBase
 
 __all__ = [
-    "CMAEvolutionStrategy",
+    "GradientOptBase",
```
Comment (itsdawei) [inline ribs/emitters/opt/__init__.py:22]: I think base class should at the end of the list as in other `__init__` files.

File: ribs/emitters/_gradient_arborescence_emitter.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ class GradientArborescenceEmitter(EmitterBase):
             :class:`ribs.archives.GridArchive`.
         x0 (np.ndarray): Initial solution.
         sigma0 (float): Initial step size / standard deviation.
-        step_size (float): Step size for the gradient optimizer
+        step_size (float): Step size for the gradient optimizer.
```
Comment (itsdawei) [inline ribs/emitters/_gradient_arborescence_emitter.py:35]: Do we want this to be a kwargs to the optimizers?

File: ribs/emitters/opt/_gradient_ascent_opt.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+class GradientAscentOpt(GradientOptBase):
+    """Vanilla gradient ascent.
+
+    Args:
+        theta0 (array-like): 1D initial solution.
+        step_size (float): Used to scale the gradient during the update.
+    """
+
+    def __init__(self, theta0, step_size):
```
Comment (itsdawei) [inline ribs/emitters/opt/_gradient_ascent_opt.py:20]: can we have a default for step_size like adam_opt?

File: ribs/emitters/opt/_gradient_opt_base.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+"""Provides GradientOptBase."""
+from abc import ABC, abstractmethod
+
+
+class GradientOptBase(ABC):
```
Comment (itsdawei) [inline ribs/emitters/opt/_gradient_opt_base.py:5]: We might be able to add a constructor to initialize step_size? All gradient opt needs it right?

File: ribs/emitters/opt/__init__.py (hunk 1 above)
Comment (btjanaka) [reply ribs/emitters/opt/__init__.py]: done

File: ribs/emitters/_gradient_arborescence_emitter.py (hunk 2 above)
Comment (btjanaka) [reply ribs/emitters/_gradient_arborescence_emitter.py:35]: All optimizers use it and it is a commonly set parameter, so let's keep it

File: ribs/emitters/opt/_gradient_ascent_opt.py (hunk 3 above)
Comment (btjanaka) [reply ribs/emitters/opt/_gradient_ascent_opt.py]: resolved offline - gradient ascent usually does not have a default step_size as in pytorch

**Review by itsdawei — APPROVED** [review]

**Pull request merged**